### PR TITLE
Optimize outer joins with one or more empty sources 

### PR DESCRIPTION
**Pull request by ahmadghazal64**
Address the outer join case for #15412

== NO RELEASE NOTE ==

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java [hunk 1]
```diff
@@ -15,21 +15,43 @@
...
+    // Build assignment list for the new Project node as: X=X if X is from non-empty child and X=null otherwise.
+    public static Assignments buildAssignments(Collection<VariableReferenceExpression> variables, PlanNode nonEmptyChild)
+    {
+        Assignments.Builder builder = Assignments.builder();
+        variables.forEach(variable -> {
+            if (nonEmptyChild.getOutputVariables().contains(variable)) {
+                builder.put(variable, variable);
+            }
+            else {
+                builder.put(variable, constantNull(variable.getType()));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java:49]: What happens if there was a const/literal on the empty side and may be we projected it out? like:

SELECT T1.*, T2.* FROM T1 LEFT OUTER JOIN (SELECT 1 AS x, .. FROM T2 LIMIT 0) USING(key)

Good to add this as a test case.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java (hunk 1 above)
Comment (ahmadghazal64) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java:49]: Good to add but the result should be null in this case too.

Comment (ahmadghazal64) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java:49]: Added more tests that covers this case.

**Review by kaikalur — APPROVED** [review]
LGTM

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -167,22 +168,59 @@ public void selectLargeInterval()
...
+                .setSystemProperty(OPTIMIZE_JOINS_WITH_EMPTY_SOURCES, "true")
+                .build();
+        Session sessionWithoutEmptyJoin = Session.builder(getSession())
+                .setSystemProperty(OPTIMIZE_JOINS_WITH_EMPTY_SOURCES, "false")
+                .build();
+        emptyJoinQueries(sessionWithEmptyJoin);
+        emptyJoinQueries(sessionWithoutEmptyJoin);
+    }
+
+    private void emptyJoinQueries(Session session)
```
[the remark this answers is not in the document]
Comment (ahmadghazal64) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:181]: Done and PR updated.

[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:181]: I still don't see the right outer join test here.  Am i missing it?

[the remark this answers is not in the document]
Comment (ahmadghazal64) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:181]: My fault and my push did not include the latest changes. It is done now and you can see it below line 215-225

**Review by rschlussel — APPROVED** [review]

**Review by rongrong — APPROVED** [review]
Just a few nits. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java [hunk 3]
```diff
@@ -48,11 +70,44 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
         leftChildEmpty = isAtMost(context.getLookup().resolve(joinNode.getLeft()), context.getLookup(), 0);
         rightChildEmpty = isAtMost(context.getLookup().resolve(joinNode.getRight()), context.getLookup(), 0);
 
-        if ((!leftChildEmpty && !rightChildEmpty) || joinNode.getType() != JoinNode.Type.INNER) {
-            return Result.empty();
+        /*
+        Prune joins with one or empty sources in the following cases. The pruning is done by replacing the whole join by empty values node.
```
[the remark this answers is not in the document]
Comment (ahmadghazal64) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java [hunk 4]
```diff
@@ -15,21 +15,43 @@
...
 public class EliminateEmptyJoins
         implements Rule<JoinNode>
 {
     private static final Pattern<JoinNode> PATTERN = join();
 
+    // Build assignment list for the new Project node as: X=X if X is from non-empty child and X=null otherwise.
+    public static Assignments buildAssignments(Collection<VariableReferenceExpression> variables, PlanNode nonEmptyChild)
+    {
+        Assignments.Builder builder = Assignments.builder();
+        variables.forEach(variable -> {
```
[the remark this answers is not in the document]
Comment (ahmadghazal64) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateEmptyJoins.java]: Done

**Pull request merged**